Feature/api accrual

a2sh3r opened the pull request.

a2sh3r committed:
Created base service

a2sh3r committed:
Created register and login endpoints

a2sh3r committed:
User service

a2sh3r committed:
Orders api

a2sh3r committed:
Balance api

a2sh3r committed:
GET /api/user/withdrawals

a2sh3r committed:
Accrual client + tests

a2sh3r committed:
Tests

a2sh3r committed:
Tests

a2sh3r committed:
Merge branch 'master' into feature/api-accrual

a2sh3r committed:
Tests

a2sh3r committed:
Tests

a2sh3r committed:
Tests

a2sh3r committed:
Tests

a2sh3r committed:
Added more tests + fixed luhn

a2sh3r committed:
Fixed static tests

ekubyshin commented on cmd/gophermart/main.go:
```diff
@@ -1,3 +1,31 @@
+		panic(err)
+	}
+
+	if err := newApp.Run(ctx); err != nil {
```
хорошее решение с контекстом

ekubyshin commented on internal/database/db.go:
```diff
@@ -0,0 +1,66 @@
+	return db, nil
+}
+
+func runMigrations(db *sql.DB) error {
```
не понял предназначение этих миграций, если в проекте есть либа?

ekubyshin commented on internal/repository/balance_repository.go:
```diff
@@ -0,0 +1,124 @@
+	}()
+
+	_, err = tx.ExecContext(ctx, `
+		INSERT INTO withdrawals (order_number, sum, processed_at, user_id)
```
а что тут заинлайнил sql? везде вынес в переменные, а тут инлайн

ekubyshin commented on internal/utils/luhn.go:
стоит избегать таких названий пакетов, тк подобные названия не отражают, что он в себе скрывает. В го принято называть пакеты и файлы, чтобы названия отражали предназначение. В данном варианте не понятно, что здесь лежит. Со временем в проектах подобные пакеты превращаются в свалку.

a2sh3r committed:
Review fix
- changed luhn -> luhn_validation
- changed migration logic
- uninlined sql query
